src/liballoc/collections/btree/map.rs

```diff
@@ -1987,9 +1987,9 @@ where
                 }
                 (i, false) => i,
             },
-            (_, Unbounded) => max_node.keys().len(),
+            (_, Unbounded) => max_node.len(),
```

Comment:
I am not confident that this change is correct. In particular, it feels like when we first enter the loop with `BTreeMap::range(..)` on an empty BTreeMap I believe `root2` (and as such `max_node`) will be the shared root.

Can you verify that we have a test for that, and ideally also explain why it's acceptable to make this change (just in the discussion here is fine)?

I believe the other replacements that avoid calling keys() are fine, as both are when we are in `max_found` which I think means that we are currently looking at a node which has the key (i.e., won't recurse further).

I would like to see a comment somewhere in here that says that len() is not safe to call on a shared root though.

Reply:
I don't understand your concern. It's all quite the opposite. `len()` _is_ safe to call on a shared root, on every node. `len()` is even implemented without any `unsafe`, a rarity in the land of the btree.

In my mind, it was always a little silly to create a slice (composed from the length in the node header and some start pointer) just to get the length, while there is direct access to that length. Now that `.keys()` no longer works on all nodes, it's time to stop the silliness.

I did verify complete coverage while crafting tests for this in #68154.

You have a point that it's not necessary to shortcut the other two "silly" `keys()` calls because shared roots can't wiggle their way into those places, but then again, why prefer unsafe code over shorter safe code.

Reply:
Hm, maybe I misinterpreted the commentary in the (old) into_key_slice. It seems to suggest that calling `as_header()` is unsafe on a shared root, but I see now that doesn't make sense.

Yeah, calling len() directly here then seems fine. Resolving.